### DISCUSSION TITLE
Rnode on Helteck T114

### DISCUSSION THREAD

**Discussion by Slydew** [markqvist/Reticulum#566] [Help & Questions]
Hello,

Does Rnode Support helteck T114 https://heltec.org/project/mesh-node-t114/ ? 

nRF52840 and SX1262, supports LoRa and Bluetooth 5.0, optional 1.14 inch TFT display and GPS module as accessories.

Regards

**Comment by Slydew** [comment]
It seems similar than RAK4631 (Nordic nRF52840)

**Comment by liamcottle** [comment]
Currently not supported at this stage. I had a couple of them turn up the other day, but I haven't looked into adding support or even using them yet as I believe there's a hardware issue with these devices. A hardware revision is in the works by Heltec.

**Comment by mofosyne** [comment]
Should be fixed by now. How is it going? Does it now support the new LCD screen?

**Reply by Gitotiehub** [reply]
It works great with or without lcd screen.

**Reply by cobraPA** [reply]
It's in Mark's repo now: https://github.com/markqvist/RNode_Firmware/blob/master/Boards.h.  I didn't check CE.

**Comment by nicebread** [comment]
I tried to flash my Heltec T114, but got this error:

```
------------------------------------------------------------------------------
                               Installer Ready

Ok, that should be all the information we need. Please confirm the following
summary before proceeding. In the next step, the device will be flashed and
provisioned, so make sure that you are satisfied with your choices.

Serial port     : /dev/cu.usbmodem1101
Device type     : Heltec Mesh Node T114 850 - 950 MHz
Platform        : NRF52
Device MCU      : Nordic Semiconductor nRF52840
Firmware file   : rnode_firmware_heltec_t114.zip

------------------------------------------------------------------------------

Is the above correct? [y/N] y
[06:30:09] Checking firmware file availability...
[06:30:09] Retrieving latest version info from https://unsigned.io/firmware/latest/?v=2.4.0&variant=rnode_firmware_heltec_t114.zip
[06:30:09] The selected firmware for this board is version 1.83
[06:30:09] Using existing firmware file: rnode_firmware_heltec_t114.zip for version 1.83
[06:30:09] Verifying firmware integrity...
[06:30:10] Decompressing firmware...
[06:30:10] Firmware decompressed
[06:30:10] Flashing RNode firmware to device on /dev/cu.usbmodem1101
Upgrading target on /dev/cu.usbmodem1101 with DFU package /Users/felix/.config/rnodeconf/update/1.83/rnode_firmware_heltec_t114.zip. Flow control is disabled, Dual bank, Touch 1200
########################################
########################################
########################################
########################################
########################################
########################################
########################################
########################################
########################################
########################################
########################################
########################################
#######
Activating new firmware
Device programmed.
[06:30:39] Done flashing
[06:30:39] Waiting for NRF52 reset...
[06:30:45] Opening serial port /dev/cu.usbmodem1101...
[06:30:49] Device connected
[06:30:49] Current firmware version: 1.83
[06:30:49] Reading EEPROM...
[06:30:49] Clearing old EEPROM, this will take about 15 seconds...
[06:31:02] Waiting for NRF52 reset...
[06:31:20] Loading signing key...
[06:31:20] Bootstrapping device EEPROM...
[06:31:26] EEPROM written! Validating...
[06:31:27] Setting firmware checksum...
[06:31:34] Waiting for NRF52 reset...
[06:31:55] EEPROM was written, but validation failed. Check your settings.
```

The Heltec display shows the RNode logo and the "Missing config" message.
I am on the current `rns` version 1.0.0, and a previous `rnodeconf --eeprom-wipe` did not help either.

Any ideas how to proceed with troubleshooting?

**Reply by cobraPA** [reply]
What platform are you using?  Sometimes there are reset or serial comm issues that may look like this.  If you try after the failure with the rnodeconf --info command, how does the board look?  I know on macOS there are sometimes serial driver issues and slowing the serial rate can resolve that.  Trying another port might do it also.  What device is your 'usbmodem' ?

**Reply by nicebread** [reply]
Unfortunately not very informative:

```
~ % rnodeconf --info /dev/cu.usbmodem1101
[06:30:12] Opening serial port /dev/cu.usbmodem1101...
[06:30:15] Device connected
[06:30:15] Current firmware version: 1.83
[06:30:15] Reading EEPROM...
[06:30:15] EEPROM is invalid, no further information available
```

I am on macOS, with the CP210x VCP serial driver.
I successfully flashed a LilyGO LORA32 (but had in fact to lower the baud rate to 115200; with the default setting it gave an error).
So generally the setup does work.

I tried to lower the rate for the Heltec T114 even more (`rnodeconf --autoinstall --baud-flash 57600`), but got the same error message as before.

> Trying another port might do it also.

Tried it on one of the other USC-C ports (if that is what you meant). Now on `/dev/cu.usbmodem2101` - same issue.

**Reply by cobraPA** [reply]
That's helpful information.  There are some differences between your two targets, the nRF52 and ESP32 reset behavior can be troublesome on serial.  Do you have any USBA ports on the Mac?  I've had trouble on Windows where I could not flash on the USBC ports and have not investigated, as the USBA are just working for both target types.

Either way, you should be able to manually provision with (for a 915 MHz T114): 
rnodeconf --rom --platform 70 --product C2 --model C7 --hwrev 1 /dev/usbmoXXX
(using your port on the end, as usual)  If there is a issue with the reset state after the initial auto-install, it might need a restart in between.

If you're curious, you'll need to look at the rnodeconf source for the numbers, you can find it here: https://github.com/markqvist/Reticulum/blob/master/RNS/Utilities/rnodeconf.py#L205

**Reply by nicebread** [reply]
> **TL;DR**: I switched to a Ubuntu desktop computer with USB-A ports, everything went smooth. Didn't manage to get it to run on MacOS (details below).

On MacOS (Macbook Pro with USB-C port), it was all quite unreliable:

- Had to lower the baud rate; then flashing *sometimes* worked (doing exactly the same operation twice lead to different outcomes ...)
- Even with the lower baud rate it only worked sometimes for the LilyGO LORA32, but never (fully) for the Heltec T114 (the final step always failed). Also the manual provision failed.
- In a desperate move, I attached a USC-C to USB-A dongle, and attached the devices via the USB-A port. Didn't help.

If I should provide logs for debugging, let me know.

**Reply by cobraPA** [reply]
The programming of microcontroller devices over serial is quite a dance.  That fact that it works at all is amazing...  The only reason I was suspicious in your case is that long ago I started trying to program RNodes on an iMac (but on USBA) and failed for the serial rate problem.  I ended up modifying rnodeconf at the time, because no option existed to change serial rate.  When I got it to work and reported back to Mark, he was aware that a new driver was available to fix the issues on Mac, but Apple does not ship it.  Then the serial speed option was added and at least made for a workaround.  On your newer device with nRF52, it's a double challenge, because nRF52 seems to be even harder to get to cleanly reset over serial.  Sometimes it requires holding the boot button manually to get a clean bootloader program mode...  Anyway, glad you got it working.

USBC just seems to be even less stable.  The dance of the serial line control is timing sensitive for microcontroller reboot and entry in to program mode.  It's possible the hardware and drivers just aren't capable of getting the sequence right.  I would not expect a USBC to USBA adapter to improve that at all, as you found.

**Reply by nicebread** [reply]
Thanks for these details! In fact I would prefer to stay on my Macbook, as this is my primary device. Are there maybe alternative serial drivers that are more reliable?

**Reply by nicebread** [reply]
I also found this hardware: https://www.adafruit.com/product/5335?srsltid=AfmBOoqcEwg-mFvtvLVXxz6xmm7LvEv0IqWMnEPPKJudFjOjwEF7hYO0

But I am not sure to which pins I would connect that on the Heltec (and as it is enclosed in a box, I'd prefer to use its USB-C port)

**Reply by cobraPA** [reply]
I think the T114 already has something like that ...  But haven't tried anything on the PC/Mac side to make it (USBC) work myself.  (I don't think that is the side that needs fixing, but who knows...)

I think it should be possible to run manually, given that you said the initial flash sequence does work?  After that, it may just be an issue with what rnodeconf tries to do at the restart.  nRF52 is a little more finicky there, so it might even mean telling the user to reset and press continue, if the automated serial version doesn't come back in the correct state.
